server/types/type.go

```diff
@@ -364,6 +364,28 @@ func (t *DoltgresType) Convert(v interface{}) (interface{}, sql.ConvertInRange,
 	return nil, sql.OutOfRange, ErrUnhandledType.New(t.String(), v)
 }
 
+// GetImplicitCast is a reference to the implicit cast logic in the functions/framework package, which we can't use
+// here due to import cycles
+var GetImplicitCast func(fromType *DoltgresType, toType *DoltgresType) TypeCastFunction
+
+// GetAssignmentCast is a reference to the assignment cast logic in the functions/framework package, which we can't use
+// here due to import cycles
+var GetAssignmentCast func(fromType *DoltgresType, toType *DoltgresType) TypeCastFunction
+
+func (t *DoltgresType) ConvertToType(ctx *sql.Context, typ types.ExtendedType, val any) (any, error) {
+	dt, ok := typ.(*DoltgresType)
+	if !ok {
+		return nil, errors.Errorf("expected DoltgresType, got %T", typ)
+	}
+
+	castFn := GetAssignmentCast(dt, t)
```

Comment:
There seems to be a disparity between this and `foreignKeyComparableTypes` in the analyzer. This one operates on assignment casts, while the other function checks for an `=` operator. There will definitely be cases where `=` exists but an assignment cast does not.

Building from this, in the GMS PR we have a way to reverse the direction of foreign key value conversions (child->parent to parent->child), so we should probably validate that an assignment cast exists in both directions.

Reply:
Added the bidirectional assignment cast check here too.